tests/csapi/thread_notifications_test.go

```diff
@@ -0,0 +1,178 @@
+package csapi_tests
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/runtime"
+)
+
+func syncHasUnreadNotifs(roomID string, check func(gjson.Result, gjson.Result) bool) client.SyncCheckOpt {
+	return func(clientUserID string, topLevelSyncJSON gjson.Result) error {
+		unreadNotifications := topLevelSyncJSON.Get("rooms.join." + client.GjsonEscape(roomID) + ".unread_notifications")
+		unreadThreadNotifications := topLevelSyncJSON.Get("rooms.join." + client.GjsonEscape(roomID) + ".unread_thread_notifications")
+		if !unreadNotifications.Exists() {
+			return fmt.Errorf("syncHasUnreadNotifs(%s): missing notifications", roomID)
+		}
+		if check(unreadNotifications, unreadThreadNotifications) {
+			return nil
+		}
+		return fmt.Errorf("syncHasUnreadNotifs(%s): check function did not pass: %v / %v", roomID, unreadNotifications.Raw, unreadThreadNotifications.Raw)
+	}
+}
```

Comment:
We could craft much better errors if we changed the `check` function into a `expectedMap` and displayed a diff

```js
{
  main: {
    highlight_count: 2,
    notification_count: 4
  },
  [firstEventID]: {
    label: "firstEventIDThreadNotifications",
    highlight_count: 1,
    notification_count: 4
  }
}
```

 - `syncHasUnreadNotifs for main: Expected 2 / 4 but actual was 3  / 4 - Format is (highlights / notifications)`
 - `syncHasUnreadNotifs for firstEventIDThreadNotifications ($abcdef): Expected 1 / 2 but actual was 1 / 1 - Format is (highlights / notifications)`

Reply:
I left this somewhat generic because:

1. It seems to fit better how check functions are done with complement.
2. I suspect we will eventually want to check multiple threads.